registry/remote/repository.go

```diff
@@ -944,46 +945,99 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+	// 1. Validate Content-Type
 	mediaType, _, err := mime.ParseMediaType(resp.Header.Get("Content-Type"))
 	if err != nil {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: invalid response Content-Type: %w", resp.Request.Method, resp.Request.URL, err)
+		return ocispec.Descriptor{}, fmt.Errorf(
+			"%s %q: invalid response `Content-Type` header; %w",
+			resp.Request.Method,
+			resp.Request.URL,
+			err,
+		)
 	}
 
+	// 2. Validate Size
 	size := resp.ContentLength
 	if size == -1 {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: unknown response Content-Length", resp.Request.Method, resp.Request.URL)
-	}
-
-	// validate digest if ref is a digest
-	if refDigest, err := ref.Digest(); err == nil {
-		if err = verifyContentDigest(resp, refDigest); err != nil {
-			return ocispec.Descriptor{}, err
+		return ocispec.Descriptor{}, fmt.Errorf(
+			"%s %q: unknown response Content-Length",
+			resp.Request.Method,
+			resp.Request.URL,
+		)
+	}
+
+	// 3. Validate Digest
+	var dgst, respDigest, refDigest, hdrDigest digest.Digest
+	var refDigestErr, hdrDigestErr error
+	if refDigest, err = ref.Digest(); err != nil {
+		refDigestErr = fmt.Errorf(
+			"%s %q: invalid reference digest: `%s`; %w",
+			resp.Request.Method,
+			resp.Request.URL,
+			respDigest,
+			err,
+		)
+	}
+	if hdrDigestStr := resp.Header.Get("Docker-Content-Digest"); hdrDigestStr != "" {
+		if hdrDigest, err = digest.Parse(hdrDigestStr); err != nil {
+			hdrDigestErr = fmt.Errorf(
+				"%s %q: invalid response header value `Docker-Content-Digest`: `%s`; %w",
+				resp.Request.Method,
+				resp.Request.URL,
+				hdrDigestStr,
+				err,
+			)
 		}
-		return ocispec.Descriptor{
-			MediaType: mediaType,
-			Digest:    refDigest,
-			Size:      size,
-		}, nil
-	}
-
-	digestStr := resp.Header.Get("Docker-Content-Digest")
-	if digestStr == "" {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: empty response Docker-Content-Digest", resp.Request.Method, resp.Request.URL)
 	}
 
-	contentDigest, err := digest.Parse(digestStr)
-	if err != nil {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: invalid response Docker-Content-Digest: %s", resp.Request.Method, resp.Request.URL, digestStr)
+	switch {
+	case calculateDigest:
+		if respDigest, err = calculateDigestFromResponse(resp); err != nil {
+			return ocispec.Descriptor{}, fmt.Errorf("failed to calculate digest on response body; %w", err)
+		}
+		var digests = map[string]digest.Digest{
+			"server-supplied": hdrDigest, // via response header `Docker-Content-Digest`
+			"client-supplied": refDigest, // via request reference string
+		}
+		if err = verifyContentDigests(respDigest, digests); err != nil {
+			return ocispec.Descriptor{}, fmt.Errorf("digest validation failure; %w", err)
+		}
+		dgst = respDigest
+	case refDigestErr == nil:
+		dgst = refDigest
+	case hdrDigestErr == nil:
+		dgst = hdrDigest
```

Comment:
I'm just thinking out loud here:

1. If we need to calculate the digest
- -  Calculate the digest from the response, and compare it with the server-supplied value and client-supplied value

2. If we do not need to calculate the digest
- -  If the reference is a digest, then use it (client supplied) to generate the descriptor
- -  If the reference is not a digest, and the header contains a valid digest, then use the header digest (server supplied) to generate the descriptor
- - What if the reference is not a digest, and the header does not contain a valid digest? Will this case happen?

Reply:
If that happens, in the `oras-go` lib, the codepath [here](https://github.com/nima/oras-go/blob/issues/225/registry/remote/repository.go#L994) will finish and the struct element `Digest` will remain as `""` in the return from that function.  I think what you might be asking though is, what is the implication of that.

It's a good question, I can mock it up in the `oras` client (drop `Docker-Content-Digest` from all server responses without discrimination of the call), and see how that behaves; that might leak some clues.

Reply:
Got it.

Reply:
> What if the reference is not a digest, and the digest header is not present? Would this case happen?

If this is not a valid case and should not happen, we should just return error.

Reply:
Given that most registries return the `Docker-Content-Digest` header for `GET` requests, can the logic be simplified as following:

1. If `Docker-Content-Digest` header is present, calculating digest from content is not needed
- - If the client-supplied digest is available, verify the header digest against the client-supplied digest
- - If the client-supplied digest is not available, just respect the header digest

2. If `Docker-Content-Digest` is not present
- - If the request is a `GET`, then calculate the digest from the content. If the client-supplied digest is available, verify the content digest against the client-supplied digest; if the client-supplied digest is not available, just respect the content digest.
- - If the request is a `HEAD`: If the client-supplied digest is available, use client-supplied digest; if the client-supplied digest is not available, return error as we have no way to know the digest.

/cc @shizhMSFT What do you think?